Markdown images break the plugin invoking errors

For example:

```
![FXRMotionControllerData Blueprint representation](fxrmotioncontrollerdata-blueprint-representation.png "FXRMotionControllerData Blueprint representation")
```

Produces:

```
2024-07-18 14:00:54 [INFO] (mdbook::book): Book building has started
2024-07-18 14:00:55 [INFO] (mdbook::book): Running the epub backend
2024-07-18 14:00:55 [INFO] (mdbook::renderer): Invoking the "epub" renderer
Running mdbook-epub as plugin...
[2024-07-18T12:00:55Z ERROR mdbook_epub::generator] Failed finding/fetch resource taken from content? Look up content for possible error... Caused by: Asset was not found: 'fxrmotioncontrollerdata-blueprint-representation.png' by '/SomeProject/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-representation.png', error = No such file or directory (os error 2)
[2024-07-18T12:00:55Z ERROR mdbook_epub] Asset was not found: 'fxrmotioncontrollerdata-blueprint-representation.png' by '/SomeProject/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-representation.png', error = No such file or directory (os error 2)
2024-07-18 14:00:55 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2024-07-18 14:00:55 [ERROR] (mdbook::utils): Error: Rendering failed
2024-07-18 14:00:55 [ERROR] (mdbook::utils): 	Caused By: The "epub" renderer failed
make: *** [GNUmakefile:282: serve] Error 101
```

This however works fine with the HTML backend as intended.

Sure, I've created a new branch in our project to reproduce the issue. It can be found [here](https://dev.azure.com/SenseGlove/_git/SenseGlove-Unreal?path=%2F&version=GBdev-mdbook-epub&_a=contents). The repository is a few GBs, I've just removed everything else and kept the MDBook, so you could use the Download button from the hamburger menu.

Inside the [book.toml](https://dev.azure.com/SenseGlove/_git/SenseGlove-Unreal?path=/Handbook/book.toml&version=GBdev-mdbook-epub&_a=contents) if we comment the following config everything works as expected:

```toml
# Disabling this line everything works as expected
[output.epub]
```

But, here's the error:

```
[2024-07-19T07:54:16Z ERROR mdbook_epub::generator] Failed finding/fetch resource taken from content? Look up content for possible error... Caused by: Asset was not found: 'fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png' by '/home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png', error = No such file or directory (os error 2)
[2024-07-19T07:54:16Z ERROR mdbook_epub] Asset was not found: 'fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png' by '/home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png', error = No such file or directory (os error 2)
2024-07-19 09:54:16 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2024-07-19 09:54:16 [ERROR] (mdbook::utils): Error: Rendering failed
2024-07-19 09:54:16 [ERROR] (mdbook::utils): 	Caused By: The "epub" renderer failed
make: *** [GNUmakefile:281: serve] Error 101
```

If I remove the following lines from [this file](https://dev.azure.com/SenseGlove/_git/SenseGlove-Unreal?path=/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/README.md&version=GBdev-mdbook-epub&_a=contents) with the mdbook-epub backend enabled, everything works:

```markdown
![FXRMotionControllerData Blueprint representation](fxrmotioncontrollerdata-blueprint-representation.png "FXRMotionControllerData Blueprint representation")

![Get a joint position and rotation from FXRMotionControllerData in Blueprint](fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png "Get a joint position and rotation from FXRMotionControllerData in Blueprint")
```

As suggested by the mdbook example in their own book we keep images at the same level as the .md file since easier to address. But, for example if we move the assets to /assets/ and try to link with absolute paths it breaks the mdbook-pdf backend with images not showing up, but the PDF generation works there.

I hope that helps demonstrate the issue.

Well, I've just made the following change in our GNUmakefile

```
-       @cargo install mdbook-epub --locked
+       @cargo install --git https://github.com/Michael-F-Bryan/mdbook-epub --locked
```

And it seems v0.4.40:

```
   Compiling mdbook-epub v0.4.40 (/home/mamadou/.cargo/git/checkouts/mdbook-epub-bd1aa49df9dbc8c0/bea2901)
    Finished `release` profile [optimized] target(s) in 49.79s
   Replacing /home/mamadou/.cargo/bin/mdbook-epub
    Replaced package `mdbook-epub v0.4.40` with `mdbook-epub v0.4.40 (https://github.com/Michael-F-Bryan/mdbook-epub#bea29011)` (executable `mdbook-epub`)
Preparing the handbook source dir /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src...
Preparing the handbook target dir /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Documentation...
Preparing the handbook /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook...
Serving /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook...
2024-07-19 10:48:02 [INFO] (mdbook::book): Book building has started
2024-07-19 10:48:02 [INFO] (mdbook::book): Running the epub backend
2024-07-19 10:48:02 [INFO] (mdbook::renderer): Invoking the "epub" renderer
Running mdbook-epub as plugin...
[2024-07-19T08:48:02Z ERROR mdbook_epub::generator] Failed finding/fetch resource taken from content? Look up content for possible error... Caused by: Asset was not found: 'fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png' by '/home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png', error = No such file or directory (os error 2)
[2024-07-19T08:48:02Z ERROR mdbook_epub] Asset was not found: 'fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png' by '/home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png', error = No such file or directory (os error 2)
2024-07-19 10:48:02 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2024-07-19 10:48:02 [ERROR] (mdbook::utils): Error: Rendering failed
2024-07-19 10:48:02 [ERROR] (mdbook::utils): 	Caused By: The "epub" renderer failed
make: *** [GNUmakefile:281: serve] Error 101
```

OK, even tried this for the latest commit on the master branch:

```
-       @cargo install mdbook-epub --locked
+       @cargo install --git https://github.com/Michael-F-Bryan/mdbook-epub --rev bea29011abc66978737342320a8e539fade23413 --locked
```

```
   Compiling mdbook-epub v0.4.40 (/home/mamadou/.cargo/git/checkouts/mdbook-epub-bd1aa49df9dbc8c0/bea2901)
    Finished `release` profile [optimized] target(s) in 46.84s
   Replacing /home/mamadou/.cargo/bin/mdbook-epub                                                                 
    Replaced package `mdbook-epub v0.4.40 (https://github.com/Michael-F-Bryan/mdbook-epub#bea29011)` with `mdbook-epub v0.4.40 (https://github.com/Michael-F-Bryan/mdbook-epub?rev=bea29011abc66978737342320a8e539fade23413#bea29011)` (executable `mdbook-epub`)
Preparing the handbook source dir /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src...
Preparing the handbook target dir /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Documentation...
Preparing the handbook /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook...
Serving /home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook...
2024-07-19 10:57:50 [INFO] (mdbook::book): Book building has started
2024-07-19 10:57:50 [INFO] (mdbook::book): Running the epub backend
2024-07-19 10:57:50 [INFO] (mdbook::renderer): Invoking the "epub" renderer
Running mdbook-epub as plugin...
[2024-07-19T08:57:50Z ERROR mdbook_epub::generator] Failed finding/fetch resource taken from content? Look up content for possible error... Caused by: Asset was not found: 'fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png' by '/home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png', error = No such file or directory (os error 2)
[2024-07-19T08:57:50Z ERROR mdbook_epub] Asset was not found: 'fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png' by '/home/mamadou/dev/SGBasicDemo/Plugins/SenseGlove/Handbook/src/advanced-topics/openxr/consuming-fxrmotioncontrollerdata/index.md/fxrmotioncontrollerdata-blueprint-get-joint-position-rotation.png', error = No such file or directory (os error 2)
2024-07-19 10:57:50 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2024-07-19 10:57:50 [ERROR] (mdbook::utils): Error: Rendering failed
2024-07-19 10:57:50 [ERROR] (mdbook::utils): 	Caused By: The "epub" renderer failed
make: *** [GNUmakefile:281: serve] Error 101
```

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -1,5 +1,5 @@
-use thiserror::Error;
 use std::path::PathBuf;
+use thiserror::Error;
 
 #[derive(Error, Debug)]
 pub enum Error {
@@ -41,6 +41,9 @@ pub enum Error {
 
     #[error(transparent)]
     Io(#[from] std::io::Error),
+
+    #[error(transparent)]
+    AssetOutsideSrcDir(#[from] std::path::StripPrefixError),
 
     #[error(transparent)]
     Book(#[from] mdbook::errors::Error),

--- a/src/resources/asset.rs
+++ b/src/resources/asset.rs
@@ -82,14 +82,7 @@ impl Asset {
             return Err(Error::AssetFile(absolute_location));
         }
         // Use filename as embedded file path with content from absolute_location.
-        let binding = utils::normalize_path(Path::new(link));
-        debug!("Extracting file name from = {:?}, binding = '{binding:?}'", &full_filename.display());
-        let filename = if cfg!(target_os = "windows") {
-            binding.as_os_str().to_os_string()
-                .into_string().expect("Error getting filename for Local Asset").replace('\\', "/")
-        } else {
-            String::from(binding.as_path().to_str().unwrap())
-        };
+        let filename = full_filename.strip_prefix(src_dir)?;
 
         let asset = Asset::new(
             filename,
@@ -112,7 +105,13 @@ impl Asset {
     pub(crate) fn compute_asset_path_by_src_and_link(link: &str, chapter_dir: &PathBuf) -> PathBuf {
         let mut reassigned_asset_root: PathBuf = PathBuf::from(chapter_dir);
         let link_string = String::from(link);
-        // if chapter is a MD file, remove if from path
+        // mdbook built-in link preprocessor have `README.md` renamed and `index.md` is not exist
+        // strip the converted filename in the path 
+        if chapter_dir.ends_with("index.md") && !chapter_dir.exists() {
+            debug!("It seems a `README.md` chapter, adjust the chapter root.");
+            reassigned_asset_root.pop();
+        }
+        // if chapter is a MD file or not exist, remove if from path
         if chapter_dir.is_file() {
             reassigned_asset_root.pop();
         }

--- a/src/resources/resource.rs
+++ b/src/resources/resource.rs
@@ -48,9 +48,15 @@ pub(crate) fn find(ctx: &RenderContext) -> Result<HashMap<String, Asset>, Error>
                     continue;
                 }
                 for link in find_assets_in_markdown(&ch.content)? {
-                    let asset = match Url::parse(&link) {
-                        Ok(url) => Asset::from_url(url, &ctx.destination),
-                        Err(_) => Asset::from_local(&link, &src_dir, ch.path.as_ref().unwrap()),
+                    let asset = if let Ok(url) = Url::parse(&link) {
+                        Asset::from_url(url, &ctx.destination)
+                    } else {
+                        let result = Asset::from_local(&link, &src_dir, ch.path.as_ref().unwrap());
+                        if let Err(Error::AssetOutsideSrcDir(_)) = result {
+                            warn!("Asset '{link}' is outside source dir '{src_dir:?}' and ignored");
+                            continue;
+                        };
+                        result
                     }?;
 
                     // that is CORRECT generation way

--- a/tests/dummy/src/02_advanced/Epub_logo.svg
+++ b/tests/dummy/src/02_advanced/Epub_logo.svg
@@ -1,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
+<svg version="1.1" id="Ebene_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
+	 width="600px" height="600px" viewBox="0 0 600 600" enable-background="new 0 0 600 600" xml:space="preserve">
+<path fill="#86B918" d="M298.549,467.913L129.228,298.579L298.549,129.27l56.446,56.435L242.104,298.579l56.441,56.443
+	l169.323-169.309L320.366,38.217c-12.043-12.055-31.579-12.055-43.634,0L38.169,276.781c-12.044,12.043-12.044,31.58,0,43.633
+	l238.563,238.558c12.055,12.05,31.591,12.05,43.634,0l238.564-238.558c12.044-12.053,12.044-31.59,0-43.633L524.3,242.159
+	L298.549,467.913z"/>
+</svg>

--- a/tests/dummy/src/02_advanced/README.md
+++ b/tests/dummy/src/02_advanced/README.md
@@ -1,0 +1,5 @@
+# README.md tests
+
+## image link
+
+![Awesome image in the same folder](Epub_logo.svg "Awesome")

--- a/tests/dummy/src/SUMMARY.md
+++ b/tests/dummy/src/SUMMARY.md
@@ -2,3 +2,7 @@
 
 - [Chapter 1](./chapter_1.md)
   - [Getting started](01_getting_started/02_article.md)
+
+# Advanced
+
+- [README.md tests](./02_advanced/README.md)

--- a/tests/integration_tests.rs
+++ b/tests/integration_tests.rs
@@ -127,15 +127,36 @@ fn look_for_chapter_1_heading() {
 
 #[test]
 #[serial]
+fn look_for_chapter_2_image_link_in_readme() {
+    init_logging();
+    let mut doc = generate_epub().unwrap();
+
+    let path = if cfg!(target_os = "linux") {
+        Path::new("OEBPS").join("02_advanced").join("README.html") // linux
+    } else {
+        Path::new("OEBPS/02_advanced/README.html").to_path_buf() // windows with 'forward slash' /
+    };
+    let content = doc.0.get_resource_str_by_path(path).unwrap();
+
+    assert!(content.contains("<img src=\"Epub_logo.svg\""));
+}
+
+#[test]
+#[serial]
 fn rendered_document_contains_all_chapter_files_and_assets() {
     init_logging();
     debug!("rendered_document_contains_all_chapter_files_and_assets...");
-    let chapters = vec!["chapter_1.html", "rust-logo.png"];
+    let chapters = vec![
+        "chapter_1.html",
+        "rust-logo.png",
+        "02_advanced/README.html",
+        "02_advanced/Epub_logo.svg",
+    ];
     let mut doc = generate_epub().unwrap();
     debug!("Number of internal epub resources = {:?}", doc.0.resources);
     // number of internal epub resources for dummy test book
-    assert_eq!(10, doc.0.resources.len());
-    assert_eq!(2, doc.0.spine.len());
+    assert_eq!(12, doc.0.resources.len());
+    assert_eq!(3, doc.0.spine.len());
     assert_eq!(doc.0.mdata("title").unwrap(), "DummyBook");
     assert_eq!(doc.0.mdata("language").unwrap(), "en");
     debug!(

